src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int NATURAL_NUMBER_BORDERLINE = 0;
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double CAPACITY_MULTIPLY_BY = 1.5;
+    private static final String NEGATIVE_INDEX_ERROR = "Index should be natural number";
+    private static final String INDEX_OUT_OF_SIZE_ERROR = "Index out of list length";
+
+    private T[] data;
+    private int size;
+
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity > 0) {
+            data = (T[]) new Object[capacity];
+        } else if (capacity == 0) {
+            data = (T[]) new Object[]{};
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(NEGATIVE_INDEX_ERROR);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < data.length) {
+            data[size] = value;
+            size++;
+        } else {
+            this.ensureCapacity();
+            this.add(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        isIndexValid(index);
+        if (size < data.length) {
+            System.arraycopy(data, index, data, index + 1, size - index);
+            data[index] = value;
+            size++;
+        } else {
+            this.ensureCapacity();
+            this.add(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > data.length) {
+            this.ensureCapacity();
+            this.addAll(list);
+        }
+        for (int i = size; i < (size + list.size()); i++) {
+            data[i] = list.get(i - size);
+        }
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index + 1);
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index + 1);
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index + 1);
+        final T result = data[index];
+        if (index == size - 1) {
+            data[index] = null;
+        } else {
+            System.arraycopy(data, index + 1, data, index, size - index);
+            data[size - 1] = null;
+        }
+        size--;
+        return result;
```

Comment:
first use checkIndex instead isIndexValid() then you can not use the if condition, instead use
```System.arraycopy(data, index + 1, data, index, size - index - 1);
        data[size - 1] = null;
        size--;
        return result;

Reply:
I expected something like this
```
        checkIndex(index);
        final T removedElement = data[index];
        System.arraycopy(data, index + 1, data, index, size - index - 1);
        data[size - 1] = null;
        size--;
        return removedElement;
```
but if you don't agree, you can leave it as it was and the mentor will decide  what is right. It's ok)

Reply:
I'm really mad at myself because I tried to write something like this and it didn't work, so I added this check. And now it works